### PR TITLE
Exclude Null Value Columns in Search Response

### PR DESCRIPTION
**Pull request by sonamgupta21**
# Description
Added an `includeNulls` flag in the search response. By default, it is set to false to exclude columns with only null values from the response.

Fixes #<issue-number> (link all the GitHub issues this addresses)

# Testing
Describe how you tested this code. How can the reviewers reproduce your tests?

# Checklist:
Before marking your pull request as ready for review, complete the following.

- [ ] I have self-reviewed this PR.
- [ ] I have removed all print-debugging and commented-out code that should not be merged.
- [ ] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [ ] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

### CODE REVIEW

File: pkg/ast/pipesearch/searchHandler.go [hunk 1]
```diff
@@ -479,10 +497,30 @@ func convertRRCsToJSONResponse(rrcs []*sutils.RecordResultContainer, sizeLimit u
 		return allJsons, allCols, err
 	}
 
+	// Filter out null values from each record
+	filteredJsons := make([]map[string]interface{}, len(allJsons))
+	for i, record := range allJsons {
```
Comment (nkunal) [inline pkg/ast/pipesearch/searchHandler.go:502]: creating a copy of every record and removing the null values from the copy is going to be expensive (memory and cpu wise). 
we should just modify the allJsons map, removing things that are null

**Review by nkunal — COMMENTED** [review]
have we taken care of both the API and WS based searches  for this new flag ?

File: pkg/ast/pipesearch/searchHandler.go [hunk 2]
```diff
@@ -204,9 +205,24 @@ func ParseSearchBody(jsonSource map[string]interface{}, nowTs uint64) (string, u
...
+		includeNulls = false
+	} else {
+		switch val := includeNullsVal.(type) {
+		case bool:
+			includeNulls = val
+		case string:
+			includeNulls = val == "true"
+		default:
+			includeNulls = false
+		}
```
Comment (AndrewHess) [inline pkg/ast/pipesearch/searchHandler.go:220]: You should also have a log message here, saying we got an unexpected type, and specifying what type it is

File: pkg/ast/pipesearch/searchHandler.go [hunk 3]
```diff
@@ -482,6 +500,17 @@ func convertRRCsToJSONResponse(rrcs []*sutils.RecordResultContainer, sizeLimit u
...
 		allJsons = allJsons[:sizeLimit]
 	}
+
+	if !includeNulls {
+		for _, record := range allJsons {
+			for key, value := range record {
+				if value == nil {
+					delete(record, key)
+				}
+			}
```
Comment (AndrewHess) [inline pkg/ast/pipesearch/searchHandler.go:510]: This should work for the current pipeline, but I don't think the new pipeline calls this function (I might be wrong). Can you check?

**Review by AndrewHess — COMMENTED** [review]
Looks pretty good

File: pkg/ast/pipesearch/searchHandler.go (hunk 3 above)
Comment (sonamgupta21) [reply pkg/ast/pipesearch/searchHandler.go:510]: It does work for the new pipeline. Could you please let me know where I need to make changes for the new pipeline?

File: pkg/segment/query/processor/queryprocessor.go [hunk 4]
```diff
@@ -265,7 +267,7 @@ func (qp *QueryProcessor) GetFullResult() (*structs.PipeSearchResponseOuter, err
 // 3. Read from the update channel and the final result channel.
 //
 // Once the final result is sent, no more updates will be sent.
-func (qp *QueryProcessor) GetStreamedResult(stateChan chan *query.QueryStateChanData) error {
+func (qp *QueryProcessor) GetStreamedResult(stateChan chan *query.QueryStateChanData, includeNulls bool) error {
```
Comment (AndrewHess) [inline pkg/segment/query/processor/queryprocessor.go:270]: nit: you already added `includeNulls` as a field on `QueryProcessor`, so you probably shouldn't pass it as a parameter here; use `qp.includeNulls` instead

File: pkg/segment/query/iqr/intermediateQueryResult.go [hunk 5]
```diff
@@ -964,6 +964,9 @@ func (iqr *IQR) AsResult(qType structs.QueryType) (*structs.PipeSearchResponseOu
 	for i, record := range cValRecords {
 		recordsAsAny[i] = make(map[string]interface{})
 		for key, value := range record {
+			if !includeNulls && value.CVal == nil {
```
Comment (AndrewHess) [inline pkg/segment/query/iqr/intermediateQueryResult.go:967]: nit: 
```suggestion
			if !includeNulls && value.IsNull() {
```
This handles a few more edge cases

**Review by AndrewHess — COMMENTED** [review]
Looking good

**Review by nkunal — APPROVED** [review]

**Review by nkunal — APPROVED** [review]

**Pull request merged**